Throw Eye of Ender

Helcostr opened the pull request:
# Summary
Adds Eye of Ender throwing logic, which also includes a rudimentary structure searcher (to search for the stronghold). Begins a basic SoundEvent enum, to be grown off of the current list of sounds in the modern list.

# Description
- Added the VS2017-x64 folder to the .gitignore for easier git management (hopefully this doesn't brick any CDI pipeline)
- Punch a hole through the world to get access to `GridStructGen` to search for the "Fortress" PieceStructure.
- Fill in missing protocol for the Eye of Ender entity in 1.8
- Create the SoundEvent enums to help consolidate modern sound names into old sound identifiers (strings in 1.8)


# Added
- `Fortress.cubeset` (obtained from @NiLSPACE)
- `SoundEvent.h`
- `PSounds_1_8.cpp` (file for implementing sounds for `cProtocol_1_8_0`)
- `cThrownEnderEyeEntity`

# Todo
- Structure searcher searches a 100 block radius before reporting back all structures, instead of reporting the closest structure. Probably should optimize.
- 1.8 Minecraft separates the use item on block, and the use item commands into two separate protocol calls. Replicating the "won't throw while interacting with the `end portal frame` might be tricky in 1.8

# Related Issues
- Fixes #5348
- Fixes #3445
- Fixes #741
- Related to #5046

Helcostr committed:
Change around early return

- Move early return statement to only apply to End Portal Frame interaction (replicating minecraft functionality)

Helcostr committed:
Get Eye of Ender to show clientside

- Update CmakeLists for Entities
- Add EnderEye to ProjectileEntities list
- (TEMPORARY) Clone the functionality of ender perals to ender eyes (for testing)
- Add ender eye throw logic
- Add EnderEye Entity to Protocol 1.8
- (Tag) issue w/ 1.8 protocol double using the ender eyes.

Helcostr committed:
Initial Ender Eye code

- Convert temp Ender Pearl code to Ender Eyes code
- Add Ender Eye death (after 80 ticks) w/ sound
- Add b_SurviveAfterDeath 50/50 chance
- Add sfx on launch + death

Helcostr committed:
WIP - Converting java tick code into cpp

- No optimizations just yet.

Helcostr committed:
(Buggy) Finish porting Java code

- Rest of java code (my brain is spinning with vector math)
- "Speed" is actually the position that we need to transpose to, add player position to the LookVector (temporary testing)
- Buggy: flitting back and forth as it rises into the sky. Is it the fact that Vector3d is an object, and can't easially be copied?

Helcostr committed:
Fix flitting bug

- Speed is used SOMEWHERE to send velocity data back to the player. Set initial value to zero.
- Update Speed to send velocity data back to player (to allow for smoother interpolations) (unit per second)
- removed some useless code, added some useless code (TODO: will definitly need to clean)

Helcostr committed:
Add Stronghold to Prefabs and Generation default.

Helcostr committed:
VS2017-x64 QOL gitignore

- Instructions say to create this folder when generating a test build. so i added it to gitignore (probably temporarially)

Helcostr committed:
(WIP) Punch a hole to ask for nearest Structure

- Hardcode a bunch of names (to allow for compilation)
- Add searcher to eye of ender logic

Helcostr committed:
Debugging Time

- Silence Eye of ender debug msgs
- Check to see which Generator i am using, how far i get
- Result: Casting to GridStructGen isn't working.

Helcostr committed:
Finalize locate logic

- Finalize `cComposableGenerator` accessing `cPieceStructuresGen`
- Finish `GetNearestStructure` of `cGridStructGen`
- Modify `GetStructuresForChunk` to incoporate a +radius to search beyond the currently selected chunk
- Attach a `AString` name to `cGridStructGen` to allow for filtering of generators by name

Helcostr committed:
(WIP) structure to blockpos

- Add return same Vector3i to indicate nothing found.
- Trying Vector3i math (im getting confused)

Helcostr committed:
(Minor) Bug fix

- I doubled up on the boolean testing. That was embarrasing.

Helcostr committed:
Finish finding Stronghold

- Fix `cComposableGenerator` not properly returning the logic
- Block negative sqrDist (too big it overflows)
- Sort `NBTChunkSerializer` header
- Don't save EnderEye data (doesn't show up in 1.8.9 saves anyways)
- Add unused target variable (just incase we do want to save it, will have to rewrite code)

Helcostr committed:
Sound

- Create SoundEnum for flexability accross protocols

Helcostr committed:
Punch hole for broadcasting via SoundEvent

- `cBroadcastInterface` virtual
- `cWorld` override
- `cClientHandle` implementation

Helcostr committed:
Use SoundEvent enum

Helcostr committed:
Reorganization

- Minor: search radius is 100 chunks (mimic minecraft locate)
- Move SoundEvent to Registries
- Move Protocol related sounds to PSounds (help the file identify as a Protocol Sounds file)

MightyFilipns commented on src/WorldStorage/NBTChunkSerializer.cpp:
```diff
@@ -23,29 +23,30 @@
+#include "../Entities/Pickup.h"
+#include "../Entities/SplashPotionEntity.h"
+#include "../Entities/ThrownEnderEyeEntity.h"
+#include "../Entities/TNTEntity.h"
```
Are these include changes just sorting?

Helcostr replied:
Almost all. There is one line of `ThrownEnderEyeEntity` in there.

MightyFilipns commented on src/Protocol/Sounds/PSounds_1_8.cpp:
You should put this function into src/Protocol/Protocol/Protocol_1_8.cpp
It should be similar to the SendParticleEffect and GetProtocolParticleID.

Helcostr replied:
Discussed over discord. Agreed that it needs to be similar to `SendParticleEffect` and `GetProtocolParticleID` (splitting the two so that we can safely override `GetProtocol`, but keep the `Send` mostly unchanged, but believe that some of the data should be located in different files, to avoid trawling through mega long files. Will restructure so it's closer to the prosed strat.

Helcostr committed:
Split send pkt and evnt

- Keep the seperate files idea to help isolate enums to the file
- Have the main cpp handle the main process of the pkt forming

Helcostr committed:
Style Format (WIP)

Got interrupted, saving progress now.

Helcostr commented on src/Entities/ThrownEnderEyeEntity.cpp:
Need to rewrite into my own flavor of formula, instead of copying the HW. Already got ideas.

Helcostr committed:
Style Format (WIP)

Interrupted again

Helcostr committed:
Rip structure string

- Generators are a common filter point of a certain type of structure.
- And Style Format

Helcostr committed:
Style Format (Finish)

- Removed extra forward declaration I slipped in.
- Added extra documentation on clarity.
- Time for me to pretty the EnderEye animation code.

Helcostr reviewed (commented):
I didn' t think i could go through to make sure i was within style, but holy shit did i just did it. Over the course of... 3 days? Yay.

Helcostr committed:
Redoing code

- Remove old xRot0 and yRot0
- Replace double target w/ vector container
- Magic number identified
- Seperate end of life code into own function for easier reading.
- Replace while loop with custom code
- Move movement code to HandlePhysics
- Begin attempt of making particles (WIP 1.8)

Helcostr committed:
Finish destruction particles

- Clean up old logs
- Add custom sound (to distinguish EnderEye surviving)
- Add velocity to item spawning
- use m_ instead of a_ of world (less variable storage)

Helcostr committed:
Finalize cleanup

MightyFilipns commented on src/Entities/ThrownEnderEyeEntity.cpp:
```diff
@@ -60,22 +60,19 @@ void cThrownEnderEyeEntity::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chun
 
 void cThrownEnderEyeEntity::HandlePhysics(std::chrono::milliseconds a_Dt, cChunk & a_Chunk) {
-	static const float radToDegrees = 57.2957763671875;
+	static const float radToDegrees = 57.2957763671875;  // Store somewhere else?
```
src/Defines.h perhaps?
